planner/core/task.go

```diff
@@ -557,28 +557,190 @@ func (p *PhysicalHashJoin) attach2Task(tasks ...task) task {
 	return task
 }
 
+// for different decimal scale and precision, tiflash uses different underlying type.
+// Here we check the scale and precision to decide whether conversion is a must.
+func needDecimalConvert(tp *types.FieldType, rtp *types.FieldType) bool {
+	if tp.Decimal != rtp.Decimal {
+		return true
+	}
+	if tp.Flen >= 0 && tp.Flen <= 9 && rtp.Flen >= 0 && rtp.Flen <= 9 {
+		return false
+	}
+	if tp.Flen > 9 && tp.Flen <= 18 && rtp.Flen > 9 && rtp.Flen <= 18 {
+		return false
+	}
+	if tp.Flen > 18 && tp.Flen <= 38 && rtp.Flen > 18 && rtp.Flen <= 38 {
+		return false
+	}
+	if tp.Flen > 38 && tp.Flen <= 65 && rtp.Flen > 38 && rtp.Flen <= 65 {
+		return false
+	}
+	return true
+}
+
+func negotiateCommonType(lType, rType *types.FieldType) (*types.FieldType, bool, bool) {
+	lExtend := 0
+	rExtend := 0
+	cDec := rType.Decimal
+	if lType.Decimal < rType.Decimal {
+		lExtend = rType.Decimal - lType.Decimal
+	} else if lType.Decimal > rType.Decimal {
+		rExtend = lType.Decimal - rType.Decimal
+		cDec = lType.Decimal
+	}
+	lLen, rLen := lType.Flen+lExtend, rType.Flen+rExtend
+	cLen := lLen
+	if rLen > cLen {
+		cLen = rLen
+	}
+	if cLen > 65 {
+		cLen = 65
+	}
+	cType := types.NewFieldType(mysql.TypeNewDecimal)
+	cType.Decimal = cDec
+	cType.Flen = cLen
+	return cType, needDecimalConvert(lType, cType), needDecimalConvert(rType, cType)
+}
+
+func getProj(ctx sessionctx.Context, p PhysicalPlan) *PhysicalProjection {
+	proj := PhysicalProjection{
+		Exprs: make([]expression.Expression, 0, len(p.Schema().Columns)),
+	}.Init(ctx, p.statsInfo(), p.SelectBlockOffset())
+	for _, col := range p.Schema().Columns {
+		proj.Exprs = append(proj.Exprs, col)
+	}
+	proj.SetSchema(p.Schema().Clone())
+	proj.SetChildren(p)
+	return proj
+}
+
+func appendExpr(p *PhysicalProjection, expr expression.Expression) *expression.Column {
+	p.Exprs = append(p.Exprs, expr)
+
+	col := &expression.Column{
+		UniqueID: p.ctx.GetSessionVars().AllocPlanColumnID(),
+		RetType:  expr.GetType(),
+	}
+	col.SetCoercibility(expr.Coercibility())
+	p.schema.Append(col)
+	return col
+}
+
+// If the join key's type are decimal and needs conversion, we will add a projection below the join or exchanger if exists.
+func (p *PhysicalHashJoin) convertDecimalKeyIfNeed(lTask, rTask *mppTask) (*mppTask, *mppTask) {
+	lp := lTask.p
+	if _, ok := lp.(*PhysicalExchangeReceiver); ok {
+		lp = lp.Children()[0].Children()[0]
+	}
+	rp := rTask.p
+	if _, ok := rp.(*PhysicalExchangeReceiver); ok {
+		rp = rp.Children()[0].Children()[0]
+	}
+	// to mark if any equal cond needs to convert
+	lMask := make([]bool, len(p.EqualConditions))
+	rMask := make([]bool, len(p.EqualConditions))
+	cTypes := make([]*types.FieldType, len(p.EqualConditions))
+	lChanged := false
+	rChanged := false
+	for i, eqFunc := range p.EqualConditions {
+		lKey := eqFunc.GetArgs()[0].(*expression.Column)
```

Comment:
It seems like non-columns also need this converting, for example, where t1.deccol+1 = t2.deccol-1.

Reply:
All the expressions have been extracted as column in LogicalOptimize.